example/service/s3/getObjectWithProgress/getObjectWithProgress.go

```diff
@@ -0,0 +1,162 @@
+	"log"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	"gopkg.in/cheggaaa/pb.v1"
+)
+
+type progressWriter struct {
+	w  io.WriterAt
+	pb *pb.ProgressBar
+}
+
+func (pw *progressWriter) WriteAt(p []byte, off int64) (int, error) {
+	pw.pb.Add(len(p))
+	return pw.w.WriteAt(p, off)
+}
+
+func byteCountDecimal(b int64) string {
+	const unit = 1000
+	if b < unit {
+		return fmt.Sprintf("%d B", b)
+	}
+	div, exp := int64(unit), 0
+	for n := b / unit; n >= unit; n /= unit {
+		div *= unit
+		exp++
+	}
+	return fmt.Sprintf("%.1f %cB", float64(b)/float64(div), "kMGTPE"[exp])
+}
+
+func getFileSize(svc *s3.S3, bucket string, prefix string) (filesize int64, error error) {
+	params := &s3.ListObjectsInput {
+    Bucket: aws.String(bucket),
+		Prefix: aws.String(prefix),
+	}
+
+	resp, err := svc.ListObjects(params)
+
+	if err != nil {
+		return 0, err
+	}
+
+	return *resp.Contents[0].Size, nil
+}
+
+func validateS3URIString(s3Uri string) error {
+	hasMatch, err := regexp.MatchString("^s3://", s3Uri)
+	if err != nil {
+		return err
+	}
+	if !hasMatch {
+		return fmt.Errorf("%s not a valid S3 uri, Please enter a valid S3 uri. Ex: s3://mary/had/a/little/lamb", s3Uri)
+	}
+	return nil
+}
+
+func parseS3Uri(s3Uri string) (bucket string, prefix string) {
+	s3UriParts := strings.Split(s3Uri, "/")
+	prefix = strings.Join(s3UriParts[3:], "/")
+	bucket = s3UriParts[2]
+	return bucket, prefix
+}
+
+func parseFilename(keyString string) (filename string) {
+	ss := strings.Split(keyString, "/")
+	s := ss[len(ss)-1]
+	return s
+}
+
+func main() {
+	if len(os.Args) < 2 {
+		log.Println("USAGE ERROR: AWS_REGION=us-east-1 go run getObjWithProgress.go s3://example-s3-bucket/foo.zip")
+		return
+	}
+
+	s3Uri := os.Args[1]
+
+	err := validateS3URIString(s3Uri)
+
+	if err != nil {
+		panic(err)
+	}
+
+	bucket, key := parseS3Uri(s3Uri)
+
+	filename := parseFilename(key)
+
+	sess := session.New()
+	cred := credentials.NewChainCredentials([]credentials.Provider{
+		&credentials.EnvProvider{},
+		&credentials.SharedCredentialsProvider{},
+	})
+	auth := &aws.Config{Credentials: cred}
+
+	s3Client   := s3.New(sess, auth)
+	downloader := s3manager.NewDownloader(sess)
+
+	size, err := getFileSize(s3Client, bucket, key)
+
+	if err != nil {
+		panic(err)
+	}
+
+	log.Printf("File size is: %s", byteCountDecimal(size))
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+
+	temp, err := ioutil.TempFile(cwd, "getObjWithProgress-tmp-")
+
+	if err != nil {
+		panic(err)
+	}
+
+	defer os.Remove(temp.Name())
+	defer temp.Close()
+
+	downloadBar := pb.New64(size).SetUnits(pb.U_BYTES)
+	downloadBar.Start()
+
+	writer := &progressWriter{temp, downloadBar}
+
+	params := &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	}
+
+	if _, err := downloader.Download(writer, params); err != nil {
+		if reqErr, ok := err.(awserr.RequestFailure); ok {
+			if reqErr.StatusCode() == 304 {
```

Comment:
Could you add a comment to describe the 304 http status code case?

Reply:
Yeah this logic was left over from my CLI implementation, that checks if the file is downloaded first, removing for this example 👍